File: DataStructures/Trees/BST/BST.cpp
/*Program for a basic Binary Search Tree*/
#include<iostream>
#include<stdlib.h>
using namespace std;
class BST{
	public:
		//Structure of the New Node
		struct Node{
			int data;
			Node* left;
			Node* right;
		};
		//Inserts New Node
		struct Node* newNode(int item){
			struct Node *temp=new Node();
			temp->data=item;
			temp->left=temp->right=NULL;
			return temp;
		};
		
		//Insert Node Function
		struct Node* insert(Node* node, int data){
			/*If the tree is empty, return a new node*/
			if(node==NULL) return newNode(data);
			/*Otherwise, recur down the tree*/
			if(data<node->data)
				node->left=insert(node->left,data);
			else if(data>node->data)
				node->right=insert(node->right,data);
			return node;
		}
		
		//Delete Value
		struct Node* deleteNode(Node* node,int data){
			if(node==NULL){
				return node;
			}
			if(data<node->data){
				node->left=deleteNode(node->left,data);
			}else if(data>node->data){
				node->right=deleteNode(node->right,data);
			}
			else{
				if(node->left==NULL){
					Node* temp=node->right;
					free(node);
					return temp;
				}else if(node->right==NULL){
					Node* temp=node->left;
					free(node);
					return temp;
				}
				Node* temp=minValue(node->right);
				node->data=temp->data;
				node->right=deleteNode(node->right,temp->data);
			}
			return node;
		}
		
		//Min Value
		struct Node* minValue(Node* node){
			Node* current=node;
			while(current->left!=NULL){
				current=current->left;
			}
			return current;
		}
		
		//Prints in Inorder Method
		void inorder(Node* node){
			if(node!=NULL)
			{
				inorder(node->left);
				cout<<(node->data)<<"\t";
				inorder(node->right);
			}
		}
		
		int m(){
			Node *root=NULL;
			root=insert(root,50);
			insert(root,30);
			insert(root,20);
			insert(root,40);
			insert(root,70);
			insert(root,60);
			insert(root,80);
			cout<<"\nInorder Traversal:"<<endl;
			inorder(root);
			cout<<"\nDelete 20"<<endl;
			root=deleteNode(root,20);
			inorder(root);
			cout<<"\nDelete 30"<<endl;
			root=deleteNode(root,30);
			inorder(root);
			cout<<"\nDelete 50"<<endl;
			root=deleteNode(root,50);
			inorder(root);
			
			return 0;
		}
};
int main(){
	BST bst;
	bst.m();
}

File: DataStructures/Trees/BST/bst.py
"""
Python: Binary Search Tree (BST)
        Operations:
            i. Insert a node
           ii. Delete a node
          iii. Search a node
"""


# BST Structure
class Node:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.key = key


# In Order Traversal of BST
def inorder(root):
    if root is not None:
        inorder(root.left)
        print(root.key, end=" ")
        inorder(root.right)


# Insert Node in the tree
def insert(node, key):
    if node is None:
        return Node(key)

    if key < node.key:
        node.left = insert(node.left, key)

    else:
        node.right = insert(node.right, key)

    return node


# Returns the Minimum Value to the node
def min_value(node):
    current = node

    while current.left is not None:
        current = current.left

    return current


# For deleting a node and balance the BST
def delete_node(root, key):
    if root is None:
        return root

    if key < root.key:
        root.left = delete_node(root.left, key)

    elif key > root.key:
        root.right = delete_node(root.right, key)

    else:
        if root.left is None:
            temp = root.right
            return temp

        elif root.right is None:
            temp = root.left
            return temp

        temp = min_value(root.right)
        root.key = temp.key
        root.right = delete_node(root.right, temp.key)

    return root


# For searching a key in the tree
def search(root, key):
    if root is None:
        return False

    if root.key == key:

        return True

    if root.key > key:
        return search(root.left, key)

    else:
        return search(root.right, key)
    

if __name__ == "__main__":
    root = None
    root = insert(root, 50)
    root = insert(root, 30)
    root = insert(root, 20)
    root = insert(root, 40)
    root = insert(root, 70)
    root = insert(root, 60)
    root = insert(root, 80)

    print("In Order", end=" ")
    inorder(root)


    print("\nAfter deleting 20", end=":")
    root = delete_node(root, 20)

    inorder(root)

    print("\nAfter deleting 50", end=":")
    root = delete_node(root, 50)
    inorder(root)

    find = 80
    find_element = search(root, find)
    print("\nFinding the element {0} in the tree.".format(find))
    if find_element:
        print("{0} found.".format(find))

    else:
        print("{0} not found".format(find))
